### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Dimension): add various `surjective_injective` results

### PR DESCRIPTION
**Pull request by acmepjz**
Add various results concerning modules `M / R` and `M' / R'` with maps `i : R -> R'` and `j : M -> M'` which are compatible with scalar multiplications on `M` and `M'`.

- if `i : R' -> R` is injective and `j` is injective, then:
  - `LinearIndependent.map_of_injective_injective`: `j` preserves linear independent subsets.
  - `[lift_]rank_le_of_injective_injective`: rank of `M / R` is smaller than or equal to the rank of `M' / R'`.
- if `i` is surjective and `j` is injective, then:
  - `LinearIndependent.map_of_surjective_injective`: `j` preserves linear independent subsets.
  - `[lift_]rank_le_of_surjective_injective`: rank of `M / R` is smaller than or equal to the rank of `M' / R'`.
- if `i` and `j` are both bijective, then `[lift_]rank_eq_of_equiv_equiv`: rank of `M / R` is equal to the rank of `M' / R'`.

Also add the `Algebra` versions of these results.

---

Discussion thread: https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/.E2.9C.94.20.60E.2FL.2FF.60.20and.20.60K.2FL'.2FF.60.20isomorphic.20.3D.3E.20.60.5BE.3AL.5D.3D.5BK.3AL'.5D.60/near/408072594

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 1]
```diff
@@ -547,6 +547,117 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+/-- If `M / R` and `M' / R'` are modules, `i : R → R'` is a surjective map which maps zero to zero,
+`j : M →+ M'` is an injective monoid map, such that the scalar multiplications on `M` and `M'` are
+compatible, then the rank of `M / R` is smaller than or equal to the rank of `M' / R'`.
+As a special case, taking `R = R'` it is `LinearMap.lift_rank_le_of_injective`. -/
+theorem lift_rank_le_of_surjective_injective (j : M →+ M')
+    (hi : Surjective i) (hj : Injective j) (hc : ∀ (s : R) (x : M), j (s • x) = i s • j x) :
+    lift.{v'} (Module.rank R M) ≤ lift.{v} (Module.rank R' M') := by
+  cases subsingleton_or_nontrivial R'
+  · cases subsingleton_or_nontrivial R
+    · simp only [rank_subsingleton, lift_one, le_refl]
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Dimension.lean:567]:
```suggestion
  · nontriviality R
```

File: Mathlib/LinearAlgebra/LinearIndependent.lean [hunk 2]
```diff
@@ -275,6 +275,22 @@ theorem LinearIndependent.map' (hv : LinearIndependent R v) (f : M →ₗ[R] M')
...
 #align linear_independent.map' LinearIndependent.map'
 
+/-- If `M / R` and `M' / R'` are modules, `i : R → R'` is a surjective map which maps zero to zero,
+`j : M →+ M'` is an injective monoid map, such that the scalar multiplications on `M` and `M'` are
+compatible, then `j` sends linearly independent families of vectors to linearly independent
+families of vectors. As a special case, taking `R = R'` it is `LinearIndependent.map'`. -/
+theorem LinearIndependent.map_of_surjective_injective {R' : Type*} {M' : Type*}
+    [Semiring R'] [AddCommMonoid M'] [Module R' M'] (hv : LinearIndependent R v)
+    (i : ZeroHom R R') (j : M →+ M') (hi : Surjective i) (hj : Injective j)
+    (hc : ∀ (s : R) (x : M), j (s • x) = i s • j x) : LinearIndependent R' (j ∘ v) := by
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/LinearIndependent.lean:285]: I know this condition is similar to [SemilinearMapClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/LinearMap.html#SemilinearMapClass), but it seems the more morally correct statement is `i : ZeroHom R' R`, `Injective i` and `j (i r' • m) = r' • j m` for `r' : R` and `m : M` in this situation. And it seems strictly more general.

Comment (acmepjz) [reply Mathlib/LinearAlgebra/LinearIndependent.lean:285]: Thanks. Do you have any suggestions on the `[lift]_rank_eq` version? Currently I keep the condition of them unchanged.

Comment (alreadydone) [reply Mathlib/LinearAlgebra/LinearIndependent.lean:285]: The eq versions uses bijections so I suppose it doesn't matter which direction they're in?

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 3]
```diff
@@ -547,6 +547,156 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+  · nontriviality R
+    haveI := Module.subsingleton R' M'
+    haveI := hj.subsingleton
+    have : Module.rank R M = 0 := rank_eq_zero_iff.2 fun _ ↦ ⟨1, one_ne_zero, Subsingleton.elim _ _⟩
+    simp only [rank_subsingleton, this, lift_zero, lift_one, zero_le]
+  rw [Module.rank_def R M]
+  refine lift_iSup_le (bddAbove_range.{v, v} _) fun S ↦ ?_
+  obtain ⟨_, h⟩ := S
+  exact cardinal_lift_le_rank_of_linearIndependent (h.map_of_injective_injective i j hi
+    (fun x ↦ by rw [← _root_.map_zero (f := j)]; exact hj (a₁ := x) (a₂ := 0)) hc)
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Dimension.lean:576]: You can avoid nontrivial/subsingleton case splits by following the proof of `lift_rank_le_of_injective`:
```suggestion
  simp_rw [Module.rank, lift_iSup (bddAbove_range.{v', v'} _), lift_iSup (bddAbove_range.{v, v} _)]
  exact ciSup_mono' (bddAbove_range.{v', v} _) fun ⟨s, li⟩ ↦ ⟨⟨j '' s,
    (li.map_of_injective_injective i j hi (fun x hx ↦ hj <| by rwa [_root_.map_zero]) hc).image⟩,
      lift_mk_le'.mpr ⟨(Equiv.Set.image j s hj).toEmbedding⟩⟩
```

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 4]
```diff
@@ -547,6 +547,156 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+As a special case, taking `R = R'` it is `LinearMap.lift_rank_le_of_injective`. -/
+theorem lift_rank_le_of_surjective_injective (i : ZeroHom R R') (j : M →+ M')
+    (hi : Surjective i) (hj : Injective j) (hc : ∀ (s : R) (x : M), j (s • x) = i s • j x) :
+    lift.{v'} (Module.rank R M) ≤ lift.{v} (Module.rank R' M') := by
+  obtain ⟨i', hi'⟩ := hi.hasRightInverse
+  refine lift_rank_le_of_injective_injective i' j (fun s h ↦ ?_) hj fun s x ↦ ?_
+  · apply_fun i at h
+    rwa [hi', _root_.map_zero] at h
+  have h := hc (i' s) x
+  rwa [hi'] at h
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Dimension.lean:590]:
```suggestion
    rwa [hi', i.map_zero] at h
  rw [hc (i' s) x, hi']
```
same in `map_of_surjective_injective`

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 5]
```diff
@@ -547,6 +547,156 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+    (hi : Bijective i) (hc : ∀ (s : R) (x : M), j (s • x) = i s • j x) :
+    lift.{v'} (Module.rank R M) = lift.{v} (Module.rank R' M') := by
+  refine le_antisymm (lift_rank_le_of_surjective_injective i j hi.surjective j.injective hc) ?_
+  obtain ⟨i', h1, h2⟩ := Function.bijective_iff_has_inverse.1 hi
+  let I' : ZeroHom R' R := ⟨i', by apply_fun i using hi.injective; rw [h2, _root_.map_zero]⟩
+  refine lift_rank_le_of_surjective_injective I' j.symm h1.surjective j.symm.injective fun s x ↦ ?_
+  apply_fun j using j.injective
+  simp only [AddMonoidHom.coe_coe, AddEquiv.apply_symm_apply, hc]
+  change _ = i (i' s) • x
+  rw [h2]
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Dimension.lean:606]:
```suggestion
    lift.{v'} (Module.rank R M) = lift.{v} (Module.rank R' M') :=
  (lift_rank_le_of_surjective_injective i j hi.surjective j.injective hc).antisymm <|
    lift_rank_le_of_injective_injective i j.symm (fun x hx ↦ hi.1 <| by rwa [i.map_zero])
      j.symm.injective fun s x ↦ j.symm_apply_eq.mpr <| by erw [hc, j.apply_symm_apply]
```

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 6]
```diff
@@ -547,6 +547,156 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+theorem lift_rank_le_of_injective_injective
+    (i : R' →+* R) (j : S →+* S') (hi : Injective i) (hj : Injective j)
+    (hc : (j.comp (algebraMap R S)).comp i = algebraMap R' S') :
+    lift.{v'} (Module.rank R S) ≤ lift.{v} (Module.rank R' S') := by
+  refine _root_.lift_rank_le_of_injective_injective i j (fun s ↦ ?_) hj fun s x ↦ ?_
+  · rw [← _root_.map_zero (f := i)]
+    exact hi (a₁ := s) (a₂ := 0)
+  have := FunLike.congr_fun hc s
+  simp only [RingHom.coe_comp, comp_apply] at this
+  simp only [smul_def, AddMonoidHom.coe_coe, map_mul, ZeroHom.coe_coe, this]
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/Dimension.lean:649]:
```suggestion
  · rw [← i.map_zero]; apply hi
  have := congr($hc s) -- some elaboration magic ... two more below can be replaced 
  simp only [RingHom.coe_comp, comp_apply] at this
  simp_rw [smul_def, AddMonoidHom.coe_coe, map_mul, this]
```

**Review by alreadydone — APPROVED** [review]

**Review by alreadydone — APPROVED** [review]
Thanks!
maintainer merge

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 7]
```diff
@@ -547,6 +547,142 @@ theorem rank_pos_iff_nontrivial : 0 < Module.rank R M ↔ Nontrivial M :=
...
+variable {R : Type w} {M : Type v} [Ring R] [AddCommGroup M] [Module R M]
+  {R' : Type w'} {M' : Type v'} [Ring R'] [AddCommGroup M'] [Module R' M']
+
+/-- If `M / R` and `M' / R'` are modules, `i : R' → R` is a map which sends non-zero elements to
+non-zero elements, `j : M →+ M'` is an injective group homomorphism, such that the scalar
+multiplications on `M` and `M'` are compatible, then the rank of `M / R` is smaller than or equal to
+the rank of `M' / R'`. As a special case, taking `R = R'` it is
+`LinearMap.lift_rank_le_of_injective`. -/
+theorem lift_rank_le_of_injective_injective (i : R' → R) (j : M →+ M')
+    (hi : ∀ s, i s = 0 → s = 0) (hj : Injective j)
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Dimension.lean:563]: Stylistic question: why do you give elements of `R'` the name `s`? Isn't `r` still available?

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**